### ISSUE TITLE
E1100 for empty definition exists both in source file and in cmd

### ISSUE THREAD

**Issue by ak-ambi** [open-watcom/open-watcom-v2#306]
I have issues with compiling the newest CMake. I prepared the minimal example to present the problem:

Source:
#define DUMMY

int main()
{
    return 0;
}

Command:
wcc386 -DDUMMY cmd_define_not_identical_to_empty_define_in_code.c

Output:
Open Watcom C32 Optimizing Compiler Version 1.9
Portions Copyright (c) 1984-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
cmd_define_not_identical_to_empty_define_in_code.c(1): Error! E1100: Definition of macro 'DUMMY' not identical to previous definition
cmd_define_not_identical_to_empty_define_in_code.c: 7 lines, 0 warnings, 1 errors

**Comment by jmalak** [comment]
It is correct behaviour because command line option -DDUMMY is equivalent to -DDUMMY=1 (see OW documentation cguide).

**Comment by jmalak** [comment]
Try to use -DDUMMY=

**Comment by ak-ambi** [comment]
Is it possible to convert the error into warning? Could it possibly become the default behaviour?

**Comment by jmalak** [comment]
This OW behaviour is OK, because it is real error that can not be changed this way.
I think it is bug in CMake to rely on non-existing compilers behaviour and should be fixed in CMake.
This behaviour is common for other compilers, by example Microsoft Visual C compiler, GNU gcc etc.
All of these compilers need to use option -Dmacro= to define macro as blank.

**Comment by ak-ambi** [comment]
I agree that thos code is not compiliant with any c standard. However, both GCC and MSVC issue a warning and CMake is compiled without a problem as warning from compiler check done during configure step is silently ignored.

As they are many projects depending on -Dmacro issuing a warning only, wouldn't it be beneficial for OW to issue a warning in that case? I mean that for example there are plenty of advices to define `D__WINSOCKAPI_` and I didn't find nor any advice neither configure script with equal sign added. I think it is de facto standard to accept macro redefinition with warning only and fixing all build scripts is cumbersome.

PS: I will report the bug to CMake and curl (as the bug is in the latter project that is used by CMake during build)

**Comment by jmalak** [comment]
It is nice example if people get warning that ignore it even if it is mistake and it should be fixed.
I don't want to change this OW behaviour by default. We could define specialized compiler option or something like -wcd/wce options (disable/enable warning), but for errors.

I already sent fix to curl.

**Comment by ak-ambi** [comment]
Ok, I think I understand your point of view - ineeded compiler is allowed to react like that. I think the problem is that devs are not aware that -D flag defines macro to "1" by default.

I think such a flag is very good idea. It could be added to CFLAGS to overcome build scripts errors of various projects.

**Comment by ak-ambi** [comment]
@jmalak I found your PR: https://github.com/curl/curl/pull/1195
I already proposed to include it in CMake, because by accident I though it is already merged.

**Comment by pchapin** [comment]
The C standard uses the language "shall" when it says that if a macro is redefined it "shall" have an identical replacement list. I'm paraphrasing; the real language is in section 6.10.3p2 (C99). It does seem appropriate for a violation of a "shall" to be considered an error.

Of course, it also makes sense to follow what the rest of the community is doing, but I tend to agree with Jiri on this issue.
